nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/MapRecord.java

```diff
@@ -328,6 +328,42 @@ public void setValue(final RecordField field, final Object value) {
         }
     }
 
+    private void removeFieldFromSchema(final RecordField field) {
+        final List<RecordField> fieldsThatRemain = new ArrayList<>(schema.getFields());
+        fieldsThatRemain.remove(field);
+        schema = new SimpleRecordSchema(fieldsThatRemain);
+    }
+
+    @Override
+    public boolean remove(final RecordField field) {
+        final Optional<RecordField> existingField = resolveField(field);
+
+        if (existingField.isPresent()) {
+            removeFieldFromSchema(existingField.get());
+            values.remove(existingField.get().getFieldName());
+            return true;
+        }
+
+        return false;
+    }
+
+    @Override
+    public void regenerateSchema() {
+        final List<RecordField> schemaFields = new ArrayList<>(schema.getFieldCount());
+
+        for (RecordField schemaField : schema.getFields()) {
+            Object fieldValue = getValue(schemaField);
+            if (fieldValue != null && fieldValue instanceof Record) {
```

Comment:
I think the null check here is not necessary, instanceof returns false for nulls.